Library/Homebrew/extend/os/linux/formula.rb

```diff
@@ -23,4 +24,22 @@ def rpath
 
   sig { params(targets: T.nilable(T.any(Pathname, String))).void }
   def deuniversalize_machos(*targets); end
+
+  sig { params(spec: SoftwareSpec).void }
+  def add_global_deps_to_spec(spec)
+    @global_deps ||= {}
+    @global_deps[name] ||= begin
+      dependency_collector = spec.dependency_collector
+      related_formula_names = Set.new([
+        name,
+        *versioned_formulae_names,
+        *spec.recursive_dependencies.map(&:name),
```

Comment:
This is the opposite way round I think.

What's happening is when it checks `binutils` (which `glibc` depends on), it'll not detect it's in the glibc dependency tree as `glibc` is a dependent rather than a dependency.

I think we want this the other way round, instead of checking the formula's recursive dependencies for the presence of "glibc", we want to check whether the formula is present in glibc's formula dependency tree.

So basically this would be the `name` etc as before and the glibc_dep_if_needed would effectively do `return if (Formula["glibc"].recursive_dependencies.map(&:name) & related_formula_names).empty?`

Reply:
> The problem here of course is there's a bit of a chicken egg problem in terms of calling `recursive_dependencies` within `DependencyCollector`... We could maybe instead do a custom one-time recursive crawl of glibc dependencies (like some one-time DependencyCollector class-initialiser?) that can add to the list of formulae to avoid while it is traversing, which should solve the issue. I can try draft something tomorrow afternoon.

Yeh, this is what I'm bumping up against. If you could just push your attempt to this PR that'd be appreciated, thanks.

Reply:
If it makes this any simpler, we're going to drop `binutils` as a dependency of `glibc` because we will vendor it with the rest of the build dependencies: https://github.com/Homebrew/homebrew-core/pull/106837#issuecomment-1219829873.  We could also download `linux-headers` as a resource instead of using the versioned formula - it's just a source archive.  That would make it so that `glibc` depends on no other formulae.

Also, since `glibc` is only used in the post install of the `gcc` formulae I don't know why we'd need it to explicitly depend on `glibc` anymore.  The post-install for the `gcc` formulae is just to make sure their spec files include the right paths if the user has brewed `glibc` installed.  I also have no idea why `llvm` depends on `glibc` but I don't think it should either.  

Just so I understand, is our goal just to make sure that users who need `glibc` and `gcc@5` (soon to be `gcc@11`) don't accidentally uninstall it and break everything?

Reply:
> Also, since `glibc` is only used in the post install of the `gcc` formulae I don't know why we'd need it to explicitly depend on `glibc` anymore
>
> Just so I understand, is our goal just to make sure that users who need glibc and gcc@5 (soon to be gcc@11) don't accidentally uninstall it and break everything?

The way it worked before is that `brew` never installed glibc for you. It installed `gcc@5` for you and the dependency in there installed glibc. Once this is merged, the glibc dependency in `gcc@5` can go.